File: rust-agent/src/convex_client.rs
use crate::config::Config;
use anyhow::{anyhow, Context, Result};
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde::Deserialize;
use serde_json::{json, Value};
use std::time::Duration;

#[derive(Debug, Clone)]
pub struct ConvexClient {
    base_url: Option<String>,
    admin_key: Option<String>,
    http: reqwest::Client,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DeviceRecord {
    #[serde(rename = "_id")]
    pub id: Option<String>,
    #[serde(rename = "macAddress")]
    pub mac_address: String,
    #[serde(default = "default_device_status")]
    pub status: String,
    #[serde(rename = "pendingRegistration", default)]
    pub pending_registration: bool,
}

impl ConvexClient {
    pub fn from_config(config: &Config) -> Result<Self> {
        Self::new(
            config.convex.deployment_url.clone(),
            config.convex.admin_key.clone(),
        )
    }

    pub fn new(base_url: impl Into<String>, admin_key: Option<String>) -> Result<Self> {
        let normalized = base_url.into().trim_end_matches('/').trim().to_string();
        let base_url = if normalized.is_empty() || is_placeholder_url(&normalized) {
            None
        } else {
            Some(normalized)
        };

        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(15))
            .build()
            .context("failed building HTTP client")?;

        Ok(Self {
            base_url,
            admin_key: admin_key.filter(|k| !k.trim().is_empty()),
            http,
        })
    }

    pub fn is_configured(&self) -> bool {
        self.base_url.is_some()
    }

    pub async fn get_devices(&self) -> Result<Vec<DeviceRecord>> {
        if !self.is_configured() {
            return Ok(Vec::new());
        }

        let value = self.call("query", "devices:getDevices", json!({})).await?;
        serde_json::from_value::<Vec<DeviceRecord>>(value)
            .context("devices:getDevices did not return a valid device list")
    }

    pub async fn register_pending_device(&self, mac: &str, name: Option<&str>) -> Result<()> {
        if !self.is_configured() {
            return Ok(());
        }

        let args = json!({
            "macAddress": mac,
            "name": name.unwrap_or("")
        });
        let _ = self
            .call("mutation", "devices:registerPendingDevice", args)
            .await?;
        Ok(())
    }

    pub async fn update_device_status(&self, mac: &str, status: &str) -> Result<()> {
        if !self.is_configured() {
            return Ok(());
        }

        let args = json!({
            "macAddress": mac,
            "status": status
        });
        let _ = self
            .call("mutation", "devices:updateDeviceStatus", args)
            .await?;
        Ok(())
    }

    async fn call(&self, endpoint: &str, path: &str, args: Value) -> Result<Value> {
        let base_url = self
            .base_url
            .as_ref()
            .ok_or_else(|| anyhow!("Convex client is not configured"))?;

        let mut headers = HeaderMap::new();
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        headers.insert(
            "Convex-Client",
            HeaderValue::from_static("presence-tracker-rs-0.1.0"),
        );

        if let Some(key) = &self.admin_key {
            let auth = HeaderValue::from_str(&format!("Convex {key}"))
                .context("invalid admin key in Authorization header")?;
            headers.insert(AUTHORIZATION, auth);
        }

        let payload = json!({
            "path": path,
            "args": args,
            "format": "json"
        });

        let url = format!("{base_url}/api/{endpoint}");
        let response = self
            .http
            .post(url)
            .headers(headers)
            .json(&payload)
            .send()
            .await
            .with_context(|| format!("Convex call failed for {path}"))?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            return Err(anyhow!("Convex {endpoint} {path} failed: {status} {body}"));
        }

        let body: Value = response
            .json()
            .await
            .with_context(|| format!("invalid JSON in Convex response for {path}"))?;

        if body
            .get("status")
            .and_then(Value::as_str)
            .is_some_and(|s| s != "success")
        {
            let message = body
                .get("errorMessage")
                .and_then(Value::as_str)
                .or_else(|| body.get("error").and_then(Value::as_str))
                .unwrap_or("unknown Convex error");
            return Err(anyhow!("Convex {endpoint} {path} error: {message}"));
        }

        Ok(extract_value(body))
    }
}

fn extract_value(body: Value) -> Value {
    if let Some(value) = body.get("value") {
        value.clone()
    } else if let Some(value) = body.get("result") {
        value.clone()
    } else {
        body
    }
}

fn default_device_status() -> String {
    "absent".to_string()
}

fn is_placeholder_url(value: &str) -> bool {
    value
        .to_ascii_lowercase()
        .contains("your-convex-deployment")
}

File: presence_tracker/convex_client.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import httpx

from presence_tracker.config import Config, is_placeholder_url


@dataclass(slots=True)
class DeviceRecord:
    id: str | None
    mac_address: str
    status: str = "absent"
    pending_registration: bool = False


@dataclass(slots=True)
class BluetoothRemovalRequest:
    id: str
    mac_address: str


class ConvexClient:
    def __init__(self, base_url: str = "", admin_key: str | None = None) -> None:
        normalized = base_url.strip().rstrip("/")
        if not normalized or is_placeholder_url(normalized):
            normalized = ""
        self.base_url = normalized
        self.admin_key = admin_key.strip() if admin_key and admin_key.strip() else None
        self.http = httpx.AsyncClient(timeout=httpx.Timeout(15.0))

    @classmethod
    def from_config(cls, config: Config) -> "ConvexClient":
        return cls(config.convex.deployment_url, config.convex.admin_key)

    @property
    def is_configured(self) -> bool:
        return bool(self.base_url)

    async def close(self) -> None:
        await self.http.aclose()

    async def get_devices(self) -> list[DeviceRecord]:
        if not self.is_configured:
            return []
        value = await self.call("query", "devices:getDevicesForPresence", {})
        if not isinstance(value, list):
            raise ValueError("devices:getDevicesForPresence did not return a list")
        return [
            DeviceRecord(
                id=item.get("_id"),
                mac_address=str(item.get("macAddress", "")),
                status=str(item.get("status", "absent")),
                pending_registration=bool(item.get("pendingRegistration", False)),
            )
            for item in value
            if isinstance(item, dict)
        ]

    async def get_bluetooth_removal_requests(self) -> list[BluetoothRemovalRequest]:
        if not self.is_configured:
            return []
        value = await self.call("query", "devices:getBluetoothRemovalRequestsForPresence", {})
        if not isinstance(value, list):
            raise ValueError("devices:getBluetoothRemovalRequestsForPresence did not return a list")
        return [
            BluetoothRemovalRequest(
                id=str(item.get("_id", "")),
                mac_address=str(item.get("macAddress", "")),
            )
            for item in value
            if isinstance(item, dict) and item.get("_id")
        ]

    async def register_pending_device(
        self,
        mac_address: str,
        name: str | None = None,
        source: str | None = None,
    ) -> None:
        if not self.is_configured:
            return
        args: dict[str, Any] = {"macAddress": mac_address, "name": name or ""}
        if source:
            args["source"] = source
        await self.call("mutation", "devices:registerPendingDevice", args)

    async def acknowledge_bluetooth_removal_request(self, request_id: str, mac_address: str) -> None:
        if not self.is_configured:
            return
        await self.call(
            "mutation",
            "devices:acknowledgeBluetoothRemovalRequest",
            {"id": request_id, "macAddress": mac_address},
        )

    async def update_device_status(self, mac_address: str, status: str) -> None:
        if not self.is_configured:
            return
        await self.call(
            "mutation",
            "devices:updateDeviceStatus",
            {"macAddress": mac_address, "status": status},
        )

    async def call(self, endpoint: str, path: str, args: dict[str, Any]) -> Any:
        if not self.base_url:
            raise RuntimeError("Convex client is not configured")

        headers = {
            "Content-Type": "application/json",
            "Convex-Client": "presence-tracker-py-0.1.0",
        }
        if self.admin_key:
            headers["Authorization"] = f"Convex {self.admin_key}"

        response = await self.http.post(
            f"{self.base_url}/api/{endpoint}",
            headers=headers,
            json={"path": path, "args": args, "format": "json"},
        )
        if response.status_code < 200 or response.status_code >= 300:
            raise RuntimeError(f"Convex {endpoint} {path} failed: {response.status_code} {response.text}")

        body = response.json()
        if isinstance(body, dict) and body.get("status") not in (None, "success"):
            message = body.get("errorMessage") or body.get("error") or "unknown Convex error"
            raise RuntimeError(f"Convex {endpoint} {path} error: {message}")
        return extract_value(body)


def extract_value(body: Any) -> Any:
    if isinstance(body, dict):
        if "value" in body:
            return body["value"]
        if "result" in body:
            return body["result"]
    return body
